Tutorial update (#9940)
PrefectHQ/prefect: docs/tutorial/tasks.md

---
description: Adding tasks to our flows
tags:
    - tutorial
    - tasks
---

Prefect tasks are a fundamental component of Prefect's workflow orchestration framework. While flows provide a high-level structure for organizing and executing your code, tasks add another atomic component that can be orchestrated and observed within the context of a flow. 

!!! note "Check out the [tasks concept doc](/docs/concepts/tasks.md) for more features and detailed information"

!!! warning Task Gotchas 

>     Tasks cannot be called from other tasks directly, but they enable concurrency, allowing you to execute multiple tasks concurrently. This concurrency can greatly enhance the efficiency and performance of your workflows.

Reviewer: ```suggestion
    Tasks cannot be called from other tasks directly, but they can be executed concurrently, greatly enhancing the efficiency and performance of your workflows.
```

```
</div>

If we click the link in our terminal and follow it to Prefect Cloud, we'll see something like this:

![Tasks provide greater visibility as well as concurrency](/img/tutorial/cloud-flow-run.png)

By encapsulating a specific unit of work within a `task`, you can define its inputs, outputs, and behavior. This modular approach allows for easier management and composition of complex workflows.

For example, you can create a `task` that makes a network call to an external API and retrieves data. This `task` can then be orchestrated alongside other tasks within the flow, providing a clear structure to our workflow.

Flows and tasks share some common features. Both have metadata attributes such as name, description (which can be in full markdown format), and tags. They also support type-checked parameters, allowing you to define the expected data types of inputs and outputs. Additionally, both flows and tasks provide functionality for retries, timeouts, and other hooks to handle failure and completion events. This overlap in functionality ensures consistency and flexibility in managing your workflows.

> Some examples of task features include:

Reviewer: ```suggestion
Task features include:
```

---
description: Learn the basics of adding tasks to our flow.
tags:
    - tutorial
    - getting started
    - basics
    - tasks
    - caching
    - concurrency
    - subflows
---

## What is a task?

> A [task](/concepts/tasks/) is a Python function decorated with a `@task` decorator. Tasks represent distinct pieces of work executed within a flow. While flows provide a high-level structure for executing your code, tasks help organize your code by adding an atomic component that can be orchestrated and observed within the context of a flow.

Reviewer: ```suggestion
A [task](/concepts/tasks/) is a Python function decorated with a `@task` decorator. Tasks are atomic pieces of work that are executed independently within a flow. Tasks, and the dependencies between them, are displayed in the flow run graph, enabling you to break down a complex flow into something you can observe and understand. When a function becomes a task, it can be executed concurrently and its return value can be cached. makes it independent of the code around itWhile flows provide a high-level structure for executing your code, 
```

> A [task](/concepts/tasks/) is a Python function decorated with a `@task` decorator. Tasks are atomic pieces of work that are executed independently within a flow. Tasks, and the dependencies between them, are displayed in the flow run graph, enabling you to break down a complex flow into something you can observe and understand. When a function becomes a task, it can be executed concurrently and its return value can be cached. makes it independent of the code around itWhile flows provide a high-level structure for executing your code, 

Reviewer: Last sentence here is a fragment.

Flows and tasks share some common features:

* Both have metadata attributes such as name, description, and tags.
* Both support type-checked parameters, allowing you to define the expected data types of inputs and outputs.
* Both provide functionality for retries, timeouts, and other hooks to handle failure and completion events.

Network calls (such as our GET requests to the GitHub API) are particularly useful as tasks because they take advantage of task features such as [retries](/concepts/tasks/#retries), [caching](/concepts/tasks/#caching), and [concurrency](/concepts/task-runners/#using-a-task-runner).

!!! warning "Tasks must be called from flows"
    All tasks must be called from within a flow. Tasks may not call other tasks directly.

!!! note "When to use tasks"

>     Not all functions need be tasks. Use tasks when the features of tasks are useful.

Reviewer: ```suggestion
    Not all functions in a flow need be tasks. Use them only when their features are useful.
```

):
    repo = get_url(f"https://api.github.com/repos/{repo_name}")
    logger = get_run_logger()
    logger.info(f"PrefectHQ/prefect repository statistics 🤓:")
    logger.info(f"Stars 🌠 : {repo['stargazers_count']}")
    logger.info(f"Forks 🍴 : {repo['forks_count']}")

if __name__ == "__main__":
    get_repo_info()
```

Running the flow in your terminal will result in something like this:

<div class="terminal">
```bash
09:55:55.412 | INFO    | prefect.engine - Created flow run 'great-ammonite' for flow 'get-repo-info'
09:55:55.499 | INFO    | Flow run 'great-ammonite' - Created task run 'get_url-0' for task 'get_url'
09:55:55.500 | INFO    | Flow run 'great-ammonite' - Executing 'get_url-0' immediately...
09:55:55.825 | INFO    | Task run 'get_url-0' - Finished in state Completed()
09:55:55.827 | INFO    | Flow run 'great-ammonite' - PrefectHQ/prefect repository statistics 🤓:
09:55:55.827 | INFO    | Flow run 'great-ammonite' - Stars 🌠 : 12157
09:55:55.827 | INFO    | Flow run 'great-ammonite' - Forks 🍴 : 1251
09:55:55.849 | INFO    | Flow run 'great-ammonite' - Finished in state Completed('All states completed.')
```
</div>

## Caching

> Tasks support the ability to cache their return value. This allows you to efficiently reuse [results](/concepts/results/) of tasks that may be expensive to run with every flow run, or reuse cached results if the inputs to a task have not changed.

Reviewer: ```suggestion
Tasks support the ability to cache their return value. This allows you to efficiently reuse [results](/concepts/results/) of tasks that may be expensive to reproduce with every flow run, or reuse cached results if the inputs to a task have not changed.
```
